src/expressions/method-call-expr.md

```diff
@@ -15,22 +15,155 @@ let log_pi = pi.unwrap_or(1.0).log(2.72);
+separate lists are maintained for inherent and extension candidates. Only
+[visible] candidates are included.
+
+(For diagnostic purposes, the search may be performed slightly differently, for
+instance searching all traits not just those in scope, or also noting
+inaccessible candidates.)
+
+## Picking a method from the candidates
+
+Once the list of candidate methods is assembled, the "picking" process
+starts.
+
+Once again, the candidate types are iterated. This time, only those types
+are iterated which can be reached via the `Deref` trait or built-in derefs;
+as noted above, this may be a shorter list than those that can be reached
+using the `Receiver` trait.
+
+For each step, picking is attempted in this order:
+
+* First, a by-value method, where the `self` type precisely matches
+  * First for inherent methods
+  * Then for extension methods
+* Then, a method where `self` is received by immutable reference (`&T`)
+  * First for inherent methods
+  * Then for extension methods
+* Then, a method where `self` is received by mutable reference (`&mut T`)
+  * First for inherent methods
+  * Then for extension methods
+* Then, a method where the `self` type is a `*const T` - this is only considered
+  if the self type is `*mut T`)
+  * First for inherent methods
+  * Then for extension methods
+* And finally, a method with a `Pin` that's reborrowed, if the `pin_ergonomics`
+  feature is enabled.
+  * First for inherent methods
+  * Then for extension methods
+
+For each of those searches, if exactly one candidate is identified,
+it's picked, and the search stops. If this results in multiple possible candidates,
+then it is an error, and the receiver must be [converted][disambiguate call]
+to an appropriate receiver type to make the method call.
+
+With the example above of `SmartPtr<T>: Receiver<Target=T>`, and the receiver
+type `&SmartPtr<Foo>`, this mechanism would pick:
+
+```rust,ignore
+impl Foo {
+   fn method(self: &SmartPtr<Foo>) {}
+}
+```
 
-Then, for each candidate type `T`, search for a [visible] method with a receiver of that type in the following places:
+but would not pick
 
-1. `T`'s inherent methods (methods implemented directly on `T`).
-1. Any of the methods provided by a [visible] trait implemented by `T`.
-   If `T` is a type parameter, methods provided by trait bounds on `T` are looked up first.
-   Then all remaining methods in scope are looked up.
+```rust,ignore
+impl Foo {
+   fn method(self: &Foo) {}
+}
+```
 
-> Note: the lookup is done for each type in order, which can occasionally lead to surprising results.
+because the receiver could not be converted to `&Foo` using the `Deref` chain,
+only the `Receiver` chain.
+
+## Extra details
+
+There are a few details not considered in this overview:
+
+* The search for candidate methods will also consider searches for
+  incoherent types if `rustc_has_incoherent_inherent_impls` is active for
+  a `dyn`, struct, enum, or foreign type.
```

Comment:
Is there any way for stable Rust code to be affected by this, e.g. by how this is applied to the `Error` trait or the `CStr` type?

Reply:
Good question, this is another thing I'll need to go and figure out

Reply:
I don't think so. To use this, the unstable feature `rustc_attrs` is [necessary](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=c5f9f1ae7243f032702e3a081a6913ce). Added e17b9b6e993373ad54ce74c98b754242c80b1e6c to show that this requires an unstable feature (pending the conversation above about whether such things should be documented at all).